### PR TITLE
Rework particle + laser beam renderer

### PR DESCRIPTION
**Pull request by Yefancy**
**Contents:**
- rework the particle system
- add docs for the particle system
- add two particle demos (~beam~ + name tag)
- fix the rendering layer of the `tempered_glass_ctm`

**How to create a laser beam:**

It's really easy, just two lines of code for a very simple particle.
```java
GTLaserBeamParticle particle = new GTLaserBeamParticle(world, startPos, endPos)
    .setBody(new ResourceLocation(GTValues.MODID,"textures/fx/laser/laser.png")); // create a beam particle and set its texture.
GTParticleManager.INSTANCE.addEffect(particle); // add it to the particle manager.
```
No matter where you call it, you can easily see it be rendered in the world.

Actually, more than two lines, because you need to manage the particle lifecycle yourself. Like when to create, when to remove, data sync, or if you want to have fade-in effects, etc.

You can refer to the demo in AssemblyLine.

https://user-images.githubusercontent.com/18493855/147933099-8de120da-3ce8-4e0e-91dd-fc4779a55f9a.mp4

Another Particle Demo --- Name Tag:

https://user-images.githubusercontent.com/18493855/148082536-56f29e7e-482c-49bc-8223-b98143cfbc91.mp4

### CODE REVIEW

File: src/main/java/gregtech/client/particle/GTLaserBeamParticle.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        }
+        ITextureObject headTexture = null;
+        if (head != null) {
+            headTexture = renderEngine.getTexture(head);
+            if (headTexture == null) {
+                headTexture = new SimpleTexture(head);
+                renderEngine.loadTexture(head, headTexture);
+            }
+        }
+        float offset = - emit * (MINECRAFT.player.ticksExisted + MINECRAFT.getRenderPartialTicks());
```
[the remark this answers is not in the document]
Comment (Yefancy) [reply src/main/java/gregtech/client/particle/GTLaserBeamParticle.java:128]: ohhh... ur right.

**Comment by Yefancy** [comment]
AssemblyLine demo has been removed. 
I post its the code here for ref.
```java
    private int beamCount;
    @SideOnly(Side.CLIENT)
    private GTLaserBeamParticle[][] beamParticles;

    @Override
    public void update() {
        super.update();
        if (getRecipeMapWorkable().isWorking()) {
            int count = (int) (getRecipeMapWorkable().getProgressPercent() * getAbilities(MultiblockAbility.IMPORT_ITEMS).size()) + 1;
            if (count != beamCount) {
                beamCount = count;
                writeCustomData(GregtechDataCodes.UPDATE_PARTICLE, this::writeParticles);
            }
            return;
        }
        if (beamCount != 0) {
            beamCount = 0;
            writeCustomData(GregtechDataCodes.UPDATE_PARTICLE, this::writeParticles);
        }
    }

    @Override
    public void writeInitialSyncData(PacketBuffer buf) {
        super.writeInitialSyncData(buf);
        writeParticles(buf);
    }

    @Override
    public void receiveInitialSyncData(PacketBuffer buf) {
        super.receiveInitialSyncData(buf);
        readParticles(buf);
    }

    @Override
    public void receiveCustomData(int dataId, PacketBuffer buf) {
        if (dataId == GregtechDataCodes.UPDATE_PARTICLE) {
            readParticles(buf);
        } else {
            super.receiveCustomData(dataId, buf);
        }
    }

    @Override
    public void onRemoval() {
        super.onRemoval();
        if (getWorld().isRemote && beamParticles != null) {
            for (GTLaserBeamParticle[] particle : beamParticles) {
                if (particle[0] != null) {
                    particle[0].setExpired();
                    particle[1].setExpired();
                }
            }
        }
        beamParticles = null;
    }

    private void writeParticles(PacketBuffer buf) {
        buf.writeVarInt(beamCount);
    }

    private void readParticles(PacketBuffer buf) {
        beamCount = buf.readVarInt();
        if (beamParticles == null) {
            beamParticles = new GTLaserBeamParticle[17][2];
        }
        for (int i = 0; i < beamParticles.length; i++) {
            GTLaserBeamParticle particle = beamParticles[i][0];
            if (i < beamCount && particle == null) {
                Vector3 startPos = new Vector3().add(
                        getPos().offset(getFrontFacing().rotateY().getOpposite(), i))
                        .add(0.5, 0, 0.5);
                Vector3 endPos = startPos.copy().subtract(0, 1, 0);

                beamParticles[i][0] = new GTLaserBeamParticle(getWorld(), startPos, endPos)
                        .setBody(new ResourceLocation(GTValues.MODID,"textures/fx/laser/laser.png"))
                        .setBeamHeight(0.125f)
                        // Try commenting or adjusting on the next four lines to see what happens
//                        .setDoubleVertical(true)
                        .setHead(new ResourceLocation(GTValues.MODID,"textures/fx/laser/laser_start.png"))
                        .setHeadWidth(0.1f)
                        .setEmit(0.2f);
                beamParticles[i][0].setOnUpdate(p -> { // remove it if machine is inValid
                    if (!isValid() || getWorld().getTileEntity(getPos()) != this.getHolder()) p.setExpired();
                });

                startPos = new Vector3().add(
                        getPos().offset(getFrontFacing().rotateY().getOpposite(), i).offset(getFrontFacing().getOpposite(), 2))
                        .add(0.5, 0, 0.5);
                endPos = startPos.copy().subtract(0, 1, 0);
                beamParticles[i][1] = new GTLaserBeamParticle(getWorld(), startPos, endPos)
                        .setBody(new ResourceLocation(GTValues.MODID,"textures/fx/laser/laser.png"))
                        .setBeamHeight(0.125f)
                        // Try commenting or adjusting on the next four lines to see what happens
//                        .setDoubleVertical(true)
                        .setHead(new ResourceLocation(GTValues.MODID,"textures/fx/laser/laser_start.png"))
                        .setHeadWidth(0.1f)
                        .setEmit(0.15f);
                beamParticles[i][1].setOnUpdate(p -> { // remove it if machine is inValid
                    if (!isValid() || !GTUtility.isPosChunkLoaded(getWorld(), getPos())) p.setExpired();
                });

                // dont forget to add particles
                GTParticleManager.INSTANCE.addEffect(beamParticles[i][0], beamParticles[i][1]);
                
            } else if (i >= beamCount && particle != null) {
                particle.setExpired();
                beamParticles[i][0] = null;
                beamParticles[i][1].setExpired();
                beamParticles[i][1] = null;
            }
        }
    }
```

File: src/main/java/gregtech/api/metatileentity/MetaTileEntityHolder.java [hunk 2]
```diff
@@ -83,6 +95,7 @@ public void notifyBlockUpdate() {
     @Override
     public void readFromNBT(@Nonnull NBTTagCompound compound) {
         super.readFromNBT(compound);
+        customName = compound.getString("CustomName");
```
Comment (DStrand1) [inline src/main/java/gregtech/api/metatileentity/MetaTileEntityHolder.java:98]: Can we wrap this around a `if (compound.hasKey("CustomName"))` for existing world compatibility?

Comment (Yefancy) [reply src/main/java/gregtech/api/metatileentity/MetaTileEntityHolder.java:98]: wdym, Neither `== null` nor `isEmpty()` is considered to have no custom name. if compound has no key will return an empty string `""`

**Review by DStrand1 — APPROVED** [review]

**Pull request merged**